surfacers/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,245 @@
+			if em.Metric("success") != nil && em.Metric("total") != nil && em.Metric("failure") == nil {
+				if failure, err := calculateFailureMetric(em); err == nil {
+					em.AddMetric("failure", metrics.NewFloat(failure))
+				} else {
+					cw.l.Errorf("Error calculating failure metric: %s", err)
+				}
+			}
+
+			cw.recordEventMetrics(em)
+		}
+	}
+}
+
+// recordEventMetrics takes an EventMetric, which can contain multiple metrics of varying types, and loops through
+// each metric in the EventMetric, parsing each metric into a structure that is supported by Cloudwatch
+func (cw *CWSurfacer) recordEventMetrics(em *metrics.EventMetrics) {
+	for _, metricKey := range em.MetricsKeys() {
+
+		switch value := em.Metric(metricKey).(type) {
+		case metrics.NumValue:
+			cw.publishMetrics(cw.newCWMetricDatum(metricKey, value.Float64(), emLabelsToDimensions(em), em.Timestamp, em.LatencyUnit))
+
+		case *metrics.Map:
+			for _, mapKey := range value.Keys() {
+				dimensions := emLabelsToDimensions(em)
+				dimensions = append(dimensions, &cloudwatch.Dimension{
+					Name:  aws.String(value.MapName),
+					Value: aws.String(mapKey),
+				})
+				cw.publishMetrics(cw.newCWMetricDatum(metricKey, value.GetKey(mapKey).Float64(), dimensions, em.Timestamp, em.LatencyUnit))
+			}
+
+		case *metrics.Distribution:
+			for i, distributionBound := range value.Data().LowerBounds {
+				dimensions := append(emLabelsToDimensions(em), &cloudwatch.Dimension{
+					Name:  aws.String(distributionDimensionName),
+					Value: aws.String(strconv.FormatFloat(distributionBound, 'f', -1, 64)),
+				})
+
+				cw.publishMetrics(cw.newCWMetricDatum(metricKey, float64(value.Data().BucketCounts[i]), dimensions, em.Timestamp, em.LatencyUnit))
+			}
+		}
+	}
+}
+
+// publish the metrics to cloudwatch, using the namespace provided from configuration
+func (cw *CWSurfacer) publishMetrics(md *cloudwatch.MetricDatum) {
+	if len(cw.cwMetricDatumCache) >= 20 {
+		_, err := cw.session.PutMetricData(&cloudwatch.PutMetricDataInput{
+			Namespace:  aws.String(cw.c.GetNamespace()),
+			MetricData: cw.cwMetricDatumCache,
+		})
+
+		if err != nil {
+			cw.l.Errorf("Failed to publish metrics to cloudwatch: %s", err)
+		}
+
+		cw.cwMetricDatumCache = cw.cwMetricDatumCache[:0]
+	}
+
+	cw.cwMetricDatumCache = append(cw.cwMetricDatumCache, md)
+}
+
+// calculateFailureMetrics calculates a failure cumalative metric, from the
+// total and success metrics in the eventmetric.
+func calculateFailureMetric(em *metrics.EventMetrics) (float64, error) {
+	successMetric, totalMetric := em.Metric("success"), em.Metric("total")
+
+	success, successOK := successMetric.(metrics.NumValue)
+	total, totalOK := totalMetric.(metrics.NumValue)
+
+	if !successOK || !totalOK {
+		return 0, fmt.Errorf("unexpected error, either success or total is not a number")
+	}
+
+	failure := total.Float64() - success.Float64()
+
+	return failure, nil
+}
+
+func (s *CWSurfacer) ignoreMetric(name string) bool {
```

Comment:
Receiver "s" needs to consistent.

Reply:
Again, sorry, thats as obvious miss by me, but thanks again for catching - was copy/paste as I was trying to keep consistent with the stackdriver implementation. Apologies

Reply:
No worries.